### ISSUE TITLE
union types

### ISSUE THREAD

**Issue by gilesbradshaw** [apollographql/apollo-client#1113]
I have a union type called TestUnion:

````
# Data type and array type for data values
interface UaDataValue {
  dataType: String
  arrayType : String
  statusCode: StatusCode
}

type UaNull implements UaDataValue {
  dataType: String
  arrayType: String
  value: String
  statusCode: StatusCode
}
type UaLong implements UaDataValue {
  dataType: String
  arrayType: String
  value: Int
  statusCode: StatusCode
}
type UaInt implements UaDataValue {
  dataType: String
  arrayType: String
  value: Int
  statusCode: StatusCode
}
type UaIntArray implements UaDataValue {
  dataType: String
  arrayType: String
  value: [Int]
  statusCode: StatusCode
}

type UaFloat implements UaDataValue {
  dataType: String
  arrayType: String
  value: Float
  statusCode: StatusCode
}


union TestUnion =  UaNull | UaLong | UaInt | UaFloat | UaIntArray

type UaNode {
  id: String!
  nodeClass: NodeClassEnum
  browseName: QualifiedName
  displayName: LocalizedText
  writeMask: Int
  userWriteMask: Int
  isAbstract: Boolean
  symmetric: Boolean
  inverseName: LocalizedText
  containsNoLoops: Boolean
  eventNotifier: Int
  nodeId: ExpandedNodeId
  description: LocalizedText
  dataValue: TestUnion
  self: UaNode
}

````

I am querying this with: 
````
const MyNodeQuery = gql`query Query($id: String!) { 
  uaNode(id: $id) { 
    id
    dataValue { 
      ... on UaIntArray {intArrayValue: value, arrayType}
      ... on UaLong {longValue: value arrayType}
      ... on UaNull {nullValue: value, arrayType}
      ... on UaFloat {doubleValue: value, arrayType}
      ... on UaInt {intValue: value, arrayType}  
    } 
  } 
}`;
````

I'm getting a UaIntArray (say) from the api ie it returns only intArrayValue 

````
{  
   "data":{  
      "uaNode":{  
         "id":"ns=2;i=10931",
         "dataValue":{  
            "intArrayValue":[  
               -6234,
               -28844
            ],
            "arrayType":"Array",
            "__typename":"UaIntArray"
         },
         "__typename":"UaNode"
      }
   }
}

````

- however in the data that gets passed to my component props (react) the other properties are filled in (longvalue, nullValue, doubleValue. intValue)  

````
 "dataValue":{  
            "intArrayValue":[  
               -6234,
               -28844
            ],
            "longValue":[  
               -6234,
               -28844
            ],
           "nullValue":[  
               -6234,
               -28844
            ],
           "doubleValue":[  
               -6234,
               -28844
            ],
           "intValue":[  
               -6234,
               -28844
            ],
}
````

**Comment by helfer** [comment]
Try seeing what happens if you query for __typename as well. IIRC there should be a warning about not using `addTypename` when fragments are used. Are you not seeing that warning?

**Comment by gilesbradshaw** [comment]
Hi I don't *think* I am seeing that warning althugh not sure where i should be looking :)

I changed query to:

````
const MyNodeQuery = gql`query Query($id: String!) { 
  uaNode(id: $id) { 
    id
    dataValue { 
      __typename
      ... on UaIntArray {intArrayValue: value, arrayType}
      ... on UaLong {longValue: value arrayType}
      ... on UaNull {nullValue: value, arrayType}
      ... on UaFloat {doubleValue: value, arrayType}
      ... on UaInt {intValue: value, arrayType}  
    } 
  } 
}`;
````

but I still see the same thing - not a show stopper though

**Comment by gilesbradshaw** [comment]
Ok I think I've fixed this - will do a PR

**Comment by gilesbradshaw** [comment]
ok it's because of this 

````
// XXX here we reach an issue - we don't know if this fragment should match or not. It's either:
  // 1. A fragment on a non-matching concrete type or interface or union
  // 2. A fragment on a matching interface or union
  // If it's 1, we don't want to return anything, if it's 2 we want to match. We can't tell the
  // difference, so for now, we just do our best to resolve the fragment but turn on partial data
  context.returnPartialData = true;
  return true;
````

Is there a reason this cannot return false?

it also messes up : 
````
value {
          __typename
          ... on UaNull {nullValue: value}
          ... on UaLong {longValue: value}
          ... on UaLocalizedText  { localizedTextValue: value { text }}
          ... on UaLocalizedTextArray  { localizedTextValueArray: value { text }}
          
        }


````
the less complex items don't come through.

**Comment by helfer** [comment]
This is fixed now, see #1483.
